src/token/stream-parser.ts

```diff
@@ -16,153 +16,392 @@ import returnValueParser from './returnvalue-token-parser';
+  }
+
+  readErrorToken(): ErrorMessageToken | Promise<ErrorMessageToken> {
+    let result;
+
+    try {
+      result = errorParser(this.buffer, this.position, this.options);
+    } catch (err: any) {
+      if (err instanceof NotEnoughDataError) {
+        return this.waitForChunk().then(() => {
+          return this.readErrorToken();
+        });
+      }
+
+      throw err;
+    }
+
+    this.position = result.offset;
+    return result.value;
+  }
+
+  readDoneInProcToken(): DoneInProcToken | Promise<DoneInProcToken> {
+    let result;
+
+    try {
+      result = doneInProcParser(this.buffer, this.position, this.options);
+    } catch (err: any) {
+      if (err instanceof NotEnoughDataError) {
+        return this.waitForChunk().then(() => {
+          return this.readDoneInProcToken();
+        });
+      }
+
+      throw err;
+    }
+
+    this.position = result.offset;
+    return result.value;
+  }
+
+  readDoneProcToken(): DoneProcToken | Promise<DoneProcToken> {
+    let result;
+
+    try {
+      result = doneProcParser(this.buffer, this.position, this.options);
+    } catch (err: any) {
+      if (err instanceof NotEnoughDataError) {
+        return this.waitForChunk().then(() => {
+          return this.readDoneProcToken();
+        });
+      }
+
+      throw err;
+    }
+
+    this.position = result.offset;
+    return result.value;
+  }
+
+  readDoneToken(): DoneToken | Promise<DoneToken> {
+    let result;
+
+    try {
+      result = doneParser(this.buffer, this.position, this.options);
+    } catch (err: any) {
+      if (err instanceof NotEnoughDataError) {
+        return this.waitForChunk().then(() => {
+          return this.readDoneToken();
+        });
+      }
+
+      throw err;
+    }
+
+    this.position = result.offset;
+    return result.value;
+  }
+
+  constructor(iterable: AsyncIterable<Buffer> | Iterable<Buffer>, debug: Debug, options: ParserOptions) {
     this.debug = debug;
     this.colMetadata = [];
     this.options = options;
 
-    this.streamBuffer = streamBuffer;
-    this.suspended = false;
-    this.next = undefined;
-  }
+    this.iterator = ((iterable as AsyncIterable<Buffer>)[Symbol.asyncIterator] || (iterable as Iterable<Buffer>)[Symbol.iterator]).call(iterable);
 
-  get buffer() {
-    return this.streamBuffer.buffer;
+    this.buffer = Buffer.alloc(0);
+    this.position = 0;
   }
 
-  get position() {
-    return this.streamBuffer.position;
-  }
+  async waitForChunk() {
+    const result = await this.iterator.next();
+    if (result.done) {
+      throw new Error('unexpected end of data');
+    }
+
+    if (this.position === this.buffer.length) {
+      this.buffer = result.value;
+    } else {
+      this.buffer = Buffer.concat([this.buffer.slice(this.position), result.value]);
+    }
 
-  set position(value) {
-    this.streamBuffer.position = value;
+    this.position = 0;
   }
 
   suspend(next: () => void) {
-    this.suspended = true;
-    this.next = next;
+
   }
```

Comment:
@MichaelSun90 @mShan0 Most of the code from here up to the end of the class is unused and can be deleted now. 🙇‍♂️